### DISCUSSION TITLE
Identifying a Tanned Scan Image and Text Boundaries

### DISCUSSION THREAD

**Discussion by srini8080** [ImageMagick/ImageMagick#3843] [Digital Image Processing]
I have scan images, some of which are tanned and some of which have very good white background.

I want to identify the ones that are tanned (like the one enclosed). The brown varies from very light to dark based on the book page

Also I want to identify the boundaries of text in the page (top, left, bottom, right) so that I can crop out the unwanted boundaries of the page. Some of the scans have black page boundaries.

Thanks

![scan](https://user-images.githubusercontent.com/12046302/122629982-4a222180-d0de-11eb-9b77-622585d8afd1.jpg)

**Comment by snibgo** [comment]
For the "tanned" (non-white) appearance:
```
magick in.jpg -verbose info:
```
The output includes:
```
  Channel statistics:
    Pixels: 643000
    Red:
      min: 5  (0.0196078)
      max: 255 (1)
      mean: 216.401 (0.848631)
      standard deviation: 53.8275 (0.211088)
      kurtosis: 6.64368
      skewness: -2.84059
      entropy: 0.653251
    Green:
      min: 0  (0)
      max: 255 (1)
      mean: 204.257 (0.801008)
      standard deviation: 52.1801 (0.204628)
      kurtosis: 6.02531
      skewness: -2.7124
      entropy: 0.701759
    Blue:
      min: 0  (0)
      max: 243 (0.952941)
      mean: 183.615 (0.72006)
      standard deviation: 48.7063 (0.191005)
      kurtosis: 4.79313
      skewness: -2.43629
      entropy: 0.764535
```
So the means of the channels are very different: 0.849, 0.801 and 0.720. If you prefer a single number, you might scale to one pixel and take the Chroma channel of HCL. Windows CMD syntax:
```
magick 122629982-4a222180-d0de-11eb-9b77-622585d8afd1.jpg -scale "1x1^!" -colorspace HCL -format %[fx:mean.g] info:

0.128571
```
The result is on a scale of 0.0 meaning the image is grayscale to 1.0 meaning it is fully saturated colour. Choose your own threshold.

**Reply by srini8080** [reply]
that worked for me. I did some sampling and came up with a threshold. Thanks

Needed one other help. 

1. I have a color image on a tanned paper (attached)
2. I have a color image on a white paper (attached)

How do I distinguish if

a. The page is just BW with no color (your threshold of 0 works here)
b. The page is just tanned with no color (I have sampled it)
c. The page has white background with some color image
d. The page has tanned background with some color image

![colorimageonwhitepaper](https://user-images.githubusercontent.com/12046302/122659578-2f5fb380-d197-11eb-9e0f-ca0bc8adbfb1.jpg)

![tannedwithcolor](https://user-images.githubusercontent.com/12046302/122659567-063f2300-d197-11eb-9549-3891299c5729.jpg)

**Comment by snibgo** [comment]
When the page has colour printing, the overall average chroma tells us nothing about the paper. But we can do the same test on just the lightest pixels, assuming that those are from the paper, with no ink.

For "lightest pixels", we take the top 1% of pixels, speaking roughly. We make a mask that is white where we have we have one of those pixels, and black otherwise. Then "-compose CopyOpacity -composite" to make all the inked pixels of the original transparent. Write x1.png for debugging. "-scale" to one pixel find the average of the paper pixels, in sRGB space. We convert to HCL to find the chroma, as above.

Windows BAT script.
```
set SRC=%1

magick ^
  %SRC% ^
  ( +clone ^
    -contrast-stretch 0x1%% ^
    -fill Black +opaque White ^
+write x0.png ^
  ) ^
  -alpha off ^
  -compose CopyOpacity -composite ^
+write x1.png ^
  -scale "1x1^!" ^
  -colorspace HCL ^
  -format "%%[fx:mean.g]\n" info:
```
The two lines that write to x0.png and x1.png are purely for debugging, and can be removed.

**Reply by srini8080** [reply]
I am doing this in Linux. It is showing zero..... I think I am doing something wrong

convert 'image.jpg' \( +clone -contrast-stretch 0x1%% -fill Black \
                     +opaque White +write x0.png \) \
                    -alpha off -compose CopyOpacity -composite \
                    +write x1.png -scale 1\!x1\! \
                    -colorspace HCL -format "%[fx:mean.g]\n" info:

**Comment by snibgo** [comment]
For bash, don't double the percents, and do escape the parentheses: \\( and \\).

To detect whether the image has colour, you could use the mean chroma from HCL:
```
magick sheep.jpg -colorspace HCL -format %[fx:mean.g] info:
0.211152
```
If this is substantially above the chroma of the paper, then the image has colour.

However, if the image is mostly grayscale with a small colour patch, the mean may be low, and a maximum can be used:
```
magick sheep.jpg -colorspace HCL -format %[fx:maxima.g] info:
0.807843
```

The scan might be noisy, eg coloured dust, so we might blur the image first:
```
magick sheep.jpg -blur 0x5 -colorspace HCL -format %[fx:maxima.g] info:
0.554871
```

**Reply by srini8080** [reply]
That worked. if the mean is above 0.2 (rounded), then color. If mean is less than 0.2 and maxima is above 0.6 (rounded), then color. Otherwise Chroma. Tested on a bunch of images. Works better if the images are toned before running format.

Thanks
